### ISSUE TITLE
What's the purpose of `dashmap::mapref::one::Ref`

### ISSUE THREAD

**Issue by Frederik-Baetens** [xacrimon/dashmap#175]
I read that dashmap has some api differences to the stdlib's btreemap, but I'm wondering what the rationale behind Ref/RefMut is exactly. What's the added value of having a Ref/RefMut type, compared to the normal references that the standard library's BTreeMap returns? Does the use of this type offer any benefits to me as the user, or is this just necessitated by the implementation?

Perhaps related, the commented out part is unfortunately not equivalent to the match expression:
(the self.map is a dashmap)

```
    pub fn get(&self, key: KeyType) -> &ValueType {
        match self.map.get(&key) {
            Some(valref) => valref.value(),
            None => {
                let newval = ValueType::new();
                self.connection_map.insert(key, newval);
                //the unwrap is probably not entirely safe, but i'm using the map as an insert/read only map
                self.connection_map.get(&key).unwrap().value()
            },
        }
        //self.map.entry(key).or_insert(ValueType::new()).value()
```

Using the commented out part, gives me a "returns a reference to data owned by the current function" error. I'm not exactly certain why this is, but I would have thought that or_insert should behave the same as my match statement.

I think the difference lies in the use of the Ref type since with std's btreemap, the following is possible:

```
fn get<'a>(map: &'a mut BTreeMap<String, String>, key: String) -> &'a str {
    map.entry(key).or_insert("hi".to_string())
}
```
While this isn't possible:
```
fn getDash<'a>(map: &'a mut DashMap<String, String>, key: String) -> &'a str {
    map.entry(key).or_insert("hi".to_string()).value()
}
```

It is possible to return dashmap's ref type, but it isn't possible to return the reference obtained trough .value()

**Comment by xacrimon** [comment]
Ref wrappers are used to control internal map synchronisation. They need to exist so we can run custom code when the ref goes out of scope so that other threads may access the value as per the rwlock principle.

**Comment by Frederik-Baetens** [comment]
Thanks for the answer!

Is there a fundamental reason I can return `self.connection_map.get(&key).unwrap().value()` but not `self.map.entry(key).or_insert(ValueType::new()).value()`? I think the problem lies in that the .value refers to data contained within the .entry instead of directly to data owned by the map, but I wonder if it would be possible to convince the compiler that the value actually points to data in the map. 

The stdlib manages to return a reference obtained through an entry, and dashmap can return a .value() out of a Ref wrapper, but once the .entry gets used with dashmap, I can no longer do that. I don't really understand the fundamental reason behind why this would be impossible.

**Comment by xacrimon** [comment]
It's not doable, if you did that you could use the reference after the Ref was dropped, which means using a ref within a lock that we don't hold. It's impossible because the reference *can't* live longer than the ref. If it did, that'd be accessing data within an unlocked lock.

You can try returning the ref struct instead.

**Comment by Frederik-Baetens** [comment]
Oh, I think my confusion comes from that this is actually possible in dashmap 5.0, but I see you have yanked this version now. Did you yank it because this was mistakenly possible when it shouldn't be?

https://gitlab.com/freddyb/scuddb/-/blob/master/server/src/node/connection_pool.rs#L20 Here's an example of me extracting a &value from a ref wrapper without issue, but I assume that that shouldn't be possible?

**Comment by xacrimon** [comment]
Yep, that was a security bug.

**Comment by Frederik-Baetens** [comment]
Haha, everything makes a lot more sense now, I already thought it was weird that my match allowed me to return the reference!

Thanks for making such a cool library. I'll be using 4 now, and look forward to the fixed release of 5.0
